SSR Hydration (#2552)
yewstack/yew: website/docs/advanced-topics/server-side-rendering.md

During Hydration, components schedule 2 consecutive renders after it is
created. Any effects are called after the second render completes.
It is important to make sure that the render function of the your
component is side-effect free. It should not mutate any states or trigger
additional renders. If your component currently mutates states or triggers
additional renders, move them into an `use_effect` hook.

It's possible to use Struct Components with server-side rendering in
hydration, the view function will be called
multiple times before the rendered function will be called.
The DOM is considered as not connected until rendered function is called,
you should prevent any access to rendered nodes
until `rendered()` method is called.

## Example

```rust ,ignore
use yew::prelude::*;
use yew::Renderer;

#[function_component]
fn App() -> Html {
    html! {<div>{"Hello, World!"}</div>}
}

fn main() {
    let renderer = Renderer::<App>::new();

    // hydrates everything under body element, removes trailing
    // elements (if any).
    renderer.hydrate();
}
```

Example: [simple\_ssr](https://github.com/yewstack/yew/tree/master/examples/simple_ssr)

> Example: [ssr\_router](https://github.com/yewstack/yew/tree/master/examples/ssr_router)

Reviewer: I would bring these in line with how it's placed in other places:

```mdx
## Further reading

- The [`simple_ssr`](https://github.com/yewstack/yew/tree/master/examples/simple_ssr) example
- The [`ssr_router`](https://github.com/yewstack/yew/tree/master/examples/ssr_router) example
```

It's not blocking and can be done in a future PR

With this approach, developers can build a client-agnostic, SSR ready
application with data fetching with very little effort.

## SSR Hydration

Hydration is the process that connects a Yew application to the
server-side generated HTML file. By default, `ServerRender` prints

> hydratable string which includes additional information to facilitate hydration.

Reviewer 2: ```suggestion
hydratable html which includes additional information to facilitate hydration.
```

When the `Renderer::hydrate` method is called, instead of start rendering from

> scratch, Yew will reconcliate the Virtual DOM generated by the application to

Reviewer 2: ```suggestion
scratch, Yew will reconcile the Virtual DOM generated by the application with
```

the html string generated by the server renderer.

:::caution

To successfully hydrate an html representation created by the
`ServerRenderer`, the client must produce a Virtual DOM layout that

> exactly matches the one used for SSR including components that does not

Reviewer 2: ```suggestion
exactly matches the one used for SSR including components that do not
```

contain any elements. If you have any component that is only useful in
one implementation, you may want to use a `PhantomComponent` to fill the
position of the extra component.
:::

## Component Lifecycle during hydration

During Hydration, components are created in a different order

> (top-to-bottom for siblings). After a component is created, it will be

Reviewer 2: Is the exact order important? I don't think it should be documented in which order components are called, this is not behaviour to rely on, not even for normal components, much less during hydration. The reconciler should reserve the right to render them in any order it sees fit.

rendered at least twice before any effects are called. It is important to make

> sure your main function of the function component is side-effect free.

Reviewer 2: ```suggestion
sure the `view` function of your component is side-effect free.
```

This should be true also for struct components. Since it's in the advanced section, no reason to omit them from the discussion.

It should not mutate any states or trigger additional renders. If your
component currently mutates states or triggers additional renders, move
them into an `use_effect` hook.

:::danger Struct Components

Whilst it's possible to use Struct Components with server-side rendering in
hydration, the components will be created in a different order than they
are directly rendered on the client-side. The view function will be called
multiple times before the rendered function will be called and the
component will start receiving messages immediately after the `view()`
function is called after the first time. At that time, the DOM may not
be connected, you should prevent any access to `NodeRef` until
`rendered()` method is called.

When using SSR and Hydration, prefer function components whenever

> possible.

Reviewer 2: Most of this is true without specifically applying to hydration and falls into the "know what you're doing" category that the advanced topics section of the docs assume either way. About the comment about receiving messages, I think it's problematic. Foremost, [in the scheduler](https://github.com/yewstack/yew/blob/62d78d0641480ab54b33301efd8c02372e1ec63d/packages/yew/src/scheduler.rs#L228-L232) `first_render` are called before messages are processed. Instead of hydration circumventing this, why not just push a `first_rendered` lifecycle event when a component with `ComponentRenderState::Hydration` is rendered?

Author: > Instead of hydration circumventing this, why not just push a first_rendered lifecycle event when a component with ComponentRenderState::Hydration is rendered?

This is to make sure that Suspense boundary can receive suspension message before its "first render" so there will be no content flash.

Author: > The order shouldn't be taken as granted in either case, I think that's important to mention when you get into this.

`create()` is able to create an order dependent behaviour.

> I don't think this advice is sound, it's not technical in nature and not especially about hydration either. I could be construed as "struct components don't work as well with hydration" and that shouldn't be true, both types of components should work

This is simply to say that: If you are having issues with struct components, you should consider porting them to function component first.

Reviewer 2: > create() is able to create an order dependent behaviour.

True, but the top-to-bottom order we currently have for lists is undocumented and not guaranteed to work in the future. You shouldn't rely on the current ordering, even if it's technically possible. At least I'm not finding a spot in the doc that talks about this behaviour, I just find it weird to state this here, since it's not mentioned elsewhere afaik.

Author: Revised wording on struct component notice.
